src/main/scala/wasm/MiniWasm.scala

```diff
@@ -263,26 +263,30 @@ object Evaluator {
         eval(rest, stack, frame, kont, trail, ret)
       case Unreachable => throw Trap()
       case Block(ty, inner) =>
-        val k: Cont[Ans] = (retStack) =>
-          eval(rest, retStack.take(ty.toList.size) ++ stack, frame, kont, trail, ret)
-        // TODO: block can take inputs too
-        eval(inner, List(), frame, k, k :: trail, ret+1)
+        val (inputs, restStack) = stack.splitAt(ty.inps.size)
+        val restK: Cont[Ans] = (retStack) =>
+          eval(rest, retStack ++ restStack, frame, kont, trail, ret)
+        eval(inner, inputs, frame, restK, restK :: trail, ret + 1)
       case Loop(ty, inner) =>
         // We construct two continuations, one for the break (to the begining of the loop),
         // and one for fall-through to the next instruction following the syntactic structure
         // of the program.
-        val restK: Cont[Ans] = (retStack) => eval(rest, retStack.take(ty.toList.size) ++ stack, frame, kont, trail, ret)
-        def loop(stack: List[Value]): Ans = {
-          val k: Cont[Ans] = (retStack) => loop(retStack.take(ty.toList.size))
-          eval(inner, stack, frame, restK, k :: trail, ret+1)
+        val (inputs, restStack) = stack.splitAt(ty.inps.size)
+        val restK: Cont[Ans] = (retStack) => eval(rest, retStack ++ restStack, frame, kont, trail, ret)
+
+        def loopK(retStack: List[Value]): Ans = {
```

Comment:
In here, I think we don't need to take care about the number of inputs when entering continuation, because the wasm type checker will ensure that every time entering the loop continuation, the stack will always contain the exact number of values to be used. Am I correct?